sdk/core/Azure.Core/src/Shared/OperationInternalOfT.cs

```diff
@@ -223,27 +247,53 @@ public Response<T> WaitForCompletion(TimeSpan pollingInterval, CancellationToken
         }
 
-        /// <summary>
-        /// Sets the <see cref="OperationInternal{T}"/> state immediately.
-        /// </summary>
-        /// <param name="state">The <see cref="OperationState{T}"/> used to set <see cref="OperationInternalBase.HasCompleted"/> and other members.</param>
-        public void SetState(OperationState<T> state)
+        protected override async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
         {
-            if (state.HasCompleted && state.HasSucceeded)
+            using var asyncLock = await _stateLock.GetLockOrValueAsync(async, cancellationToken).ConfigureAwait(false);
+            if (asyncLock.HasValue)
             {
-                Value = state.Value!;
+                return GetResponseFromState(asyncLock.Value);
+            }
+
+            using var scope = CreateScope();
+            try
+            {
+                var state = await _operation.UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+                if (!state.HasCompleted)
+                {
+                    Interlocked.Exchange(ref _rawResponse, state.RawResponse);
+                    return state.RawResponse;
+                }
+
+                if (!state.HasSucceeded && state.OperationFailedException == null)
+                {
+                    state = OperationState<T>.Failure(state.RawResponse, await CreateException(async, state.RawResponse).ConfigureAwait(false));
+                }
+
+                asyncLock.SetValue(state);
+                return GetResponseFromState(state);
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
+                throw;
             }
-            ApplyStateAsync(false, state.RawResponse, state.HasCompleted, state.HasSucceeded, state.OperationFailedException, throwIfFailed: false).EnsureCompleted();
         }
 
-        protected override async ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken)
+        private static Response GetResponseFromState(OperationState<T> state)
         {
-            OperationState<T> state = await _operation.UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
-            if (state.HasCompleted && state.HasSucceeded)
+            if (state.HasSucceeded)
             {
-                Value = state.Value!;
+                return state.RawResponse;
             }
-            return await ApplyStateAsync(async, state.RawResponse, state.HasCompleted, state.HasSucceeded, state.OperationFailedException).ConfigureAwait(false);
+
+            throw state.OperationFailedException!;
+        }
+
+        private class FinalOperation : IOperation<T>
+        {
+            public ValueTask<OperationState<T>> UpdateStateAsync(bool async, CancellationToken cancellationToken)
+                => throw new NotSupportedException("Type is in final state");
```

Comment:
If it's technically possible this could throw out to user code, it should be more actionable and/or diagnosable. Perhaps:
```suggestion
                => throw new NotSupportedException("The operation has already completed");
```

Reply:
Fixed